rust/lance-linalg/src/distance/cosine.rs

```diff
@@ -125,6 +189,13 @@ fn cosine_scalar<T: Real + Sum>(x: &[T], x_norm: T, y: &[T]) -> T {
     T::one().sub(xy.div(x_norm.mul(y_sq.sqrt())))
 }
 
+#[inline]
+fn cosine_scalar_fast<T: Real + Sum>(x: &[T], x_norm: T, y: &[T], y_norm: T) -> T {
+    let xy = dot(x, y);
+    // 1 - xy / (sqrt(x_sq) * sqrt(y_sq))
+    T::one().sub(xy.div(x_norm.mul(y_norm)))
```

Comment:
Is there danger of overflow for `x_norm.mul(y_norm)` ?

Reply:
That's a good question. we can probably cast it to f64 at the last step?

Reply:
What about this?
```suggestion
    T::one().sub(xy.div(x_norm).div(y_norm))
```

Reply:
Ah, that doesn't really help.

Reply:
I upcast all of them to f64